### PR TITLE
feat(fed): offline backends - support v4 api version of list-prekeys endpoint pt.1 (AR-3171)

### PR DESCRIPTION
**Pull request by yamilmedina**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Since api v4, `/users/list-prekeys` endpoint will return a different response. This is changed to:

```
{
  "failed_to_list": [
    {
      "domain": "example.com",
      "id": "99db9768-04e3-4b5d-9268-831b6a25c4ab"
    }
  ],
  "qualified_user_client_prekeys": {
    "additionalProp1": {
      "000600d0-000b-9c1a-000d-a4130002c221": {
        "44901fb0712e588f": {
          "id": 1,
          "key": "pQABAQECoQBYIOjl7hw0D8YRNq..."
        }
      }
    },
    "additionalProp2": {
      "000600d0-000b-9c1a-000d-a4130002c221": {
        "44901fb0712e588f": {
          "id": 1,
          "key": "pQABAQECoQBYIOjl7hw0D8YRNq..."
        }
      }
    },
    "additionalProp3": {
      "000600d0-000b-9c1a-000d-a4130002c221": {
        "44901fb0712e588f": {
          "id": 1,
          "key": "pQABAQECoQBYIOjl7hw0D8YRNq..."
        }
      }
    }
  }
}
```

### Causes (Optional)

Breaks the retrocompatibility and api versioning we have on the project, since the old response until v3 is contained in this new json object.

### Solutions

- Change the return type from old `v0 -> v3` of the endpoint and only map what we need ignoring the new optional field, using `.mapSuccess` to maintain retrocompatibility.
- For v4 and up, we don't perform any transformation to handle the optional field `failed_to_list` we just delegate to ktor the mapping.

#### Test Coverage (Optional)

- [x] I have added automated test to this contribution

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [x] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/SessionEstablisher.kt [hunk 1]
```diff
@@ -43,10 +43,13 @@ internal interface SessionEstablisher {
     /**
      * Verifies if this client can send messages to all the client recipients.
      * Will fetch PreKeys and establish cryptographic sessions if needed.
+     *
+     * Since api v4 and up, returns a list of user ids [Either.Right] [List<UserId>] with clients issues
```
Comment (yamilmedina) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/SessionEstablisher.kt:47]: better docs here

**Review by typfel — CHANGES_REQUESTED** [review]

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/SessionEstablisher.kt [hunk 2]
```diff
@@ -43,10 +43,13 @@ internal interface SessionEstablisher {
...
+     * Otherwise, just returns [Either.Right] [emptyList]
      */
     suspend fun prepareRecipientsForNewOutgoingMessage(
         recipients: List<Recipient>
-    ): Either<CoreFailure, Unit>
+    ): Either<CoreFailure, List<UserId>>
```
[the remark this answers is not in the document]
Comment (yamilmedina) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/SessionEstablisher.kt:52]: Sounds like a good idea!

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSender.kt [hunk 3]
```diff
@@ -229,7 +229,9 @@ internal class MessageSenderImpl internal constructor(
             .flatMap { recipients ->
                 sessionEstablisher
                     .prepareRecipientsForNewOutgoingMessage(recipients)
-                    .map { recipients }
+                    .map { failedToListUserIds ->
+                        recipients
+                    }
                 // TODO(federation) filter clients with failed to get prekeys and add persist in db
```
[the remark this answers is not in the document]
Comment (yamilmedina) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSender.kt:235]: We can ignore it, in the #1651 this part is implemented and the TODO removed

**Review by MohamadJaara — APPROVED** [review]
lgtm

**Review by typfel — APPROVED** [review]
🌟

**Comment by yamilmedina** [comment]
I'll discard this PR in favor of https://github.com/wireapp/kalium/pull/1657.
So we have new breaking API changes asap on develop.

**Pull request closed without merging**